File: BmiCalc/app/src/main/java/pamo/bmicalc/ui/quiz/QuestionsXMLparser.java
package pamo.bmicalc.ui.quiz;

import android.util.Log;

import org.xmlpull.v1.XmlPullParser;
import org.xmlpull.v1.XmlPullParserException;
import org.xmlpull.v1.XmlPullParserFactory;

import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.List;

public class QuestionsXMLparser {
    private static final String ns = null;
    private InputStream xmlStream;
    private XmlPullParserFactory pf;
    private QuizQuestionsModel qqm;

    public QuizQuestionsModel getQqm() {
        return qqm;
    }

    public QuestionsXMLparser(InputStream xmlStream) {
        Log.v("QuestionsXMLparser", "Entered -> QuestionsXMLparser constructor.");
        this.xmlStream = xmlStream;
    }

    public void parseXML() throws XmlPullParserException, IOException {
        try {
            Log.v("QuestionsXMLparser", "Entered -> parseXML");
            pf = XmlPullParserFactory.newInstance();
            XmlPullParser xpp = pf.newPullParser();
            xpp.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false);
            xpp.setInput(xmlStream, null);
            xpp.nextTag();
            Log.v("QuestionsXMLparser", "Before processXML method.");
            processXML(xpp);
        } finally {
            Log.v("QuestionsXMLparser", "Closing InputStream.");
            xmlStream.close();
        }
    }

    private void processXML(XmlPullParser parser) throws XmlPullParserException, IOException {
        Log.v("QuestionsXMLparser", "Entered -> processXML");
        qqm = new QuizQuestionsModel();
        QuizQuestion question = null;
        List<QuizQuestion> questions = new ArrayList<>();

        parser.require(XmlPullParser.START_TAG, ns, "quiz");
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String name = parser.getName();
            if (name.equals("question")) {
                question = parseQuestion(parser);
                Log.v("QuestionsXMLparser", "Question text retrieved: " + question.getQuestionText());
            } else {
                skip(parser);
            }
            questions.add(question);
        }
        qqm.setQuestions(questions);
        Log.v("QuestionsXMLparser", "Questions parsed: " + qqm.getQuestions().size());
    }

    private QuizQuestion parseQuestion(XmlPullParser parser) throws XmlPullParserException, IOException {
        int question_id = 0;
        String question_text = "";
        List<QuizAnswer> answers = new ArrayList<>();

        parser.require(XmlPullParser.START_TAG, ns, "question");
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String name = parser.getName();
            if (name.equals("question_id")) {
                question_id = Integer.parseInt(readTagContents(parser, "question_id"));
            } else if (name.equals("question_text")) {
                question_text = readTagContents(parser, "question_text");
            } else if (name.equals("answers")) {
                answers = parseAnswers(parser);
            } else {
                skip(parser);
            }
        }
        return new QuizQuestion(question_id, question_text, answers);
    }

    private List<QuizAnswer> parseAnswers(XmlPullParser parser) throws XmlPullParserException, IOException {
        List<QuizAnswer> answers = new ArrayList<>();
        int answer_id = 0;
        String answer_text = "";
        boolean answer_valid = false;

        parser.require(XmlPullParser.START_TAG, ns, "answers");
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.getEventType() != XmlPullParser.START_TAG) {
                continue;
            }
            String name = parser.getName();
            if (name.equals("answer_id")) {
                answer_id = Integer.parseInt(readTagContents(parser, "answer_id"));
            } else if (name.equals("answer_text")) {
                answer_text = readTagContents(parser, "answer_text");
            } else if (name.equals("answer_valid")) {
                answer_valid = Boolean.parseBoolean(readTagContents(parser, "answer_text"));
            } else {
                skip(parser);
            }
            answers.add(new QuizAnswer(answer_id, answer_text, answer_valid));
        }
        return answers;
    }

    private String readTagContents(XmlPullParser parser, String tagName) throws XmlPullParserException, IOException {
        String result = "";

        parser.require(XmlPullParser.START_TAG, ns, tagName);
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.getText();
            parser.nextTag();
        }
        return result;
    }

    private void skip(XmlPullParser parser) throws XmlPullParserException, IOException {
        if (parser.getEventType() != XmlPullParser.START_TAG) {
            throw new IllegalStateException();
        }

        int depth = 1;
        while (depth != 0) {
            switch (parser.next()) {
                case XmlPullParser.END_TAG:
                    depth--;
                    break;
                case XmlPullParser.START_TAG:
                    depth++;
                    break;
            }
        }
    }
}

File: BmiCalcKotlin/app/src/main/java/pamo/bmicalc_kotlin/ui/quiz/QuestionsXMLparser.kt
package pamo.bmicalc_kotlin.ui.quiz

import android.util.Log
import org.xmlpull.v1.XmlPullParser
import org.xmlpull.v1.XmlPullParserException
import org.xmlpull.v1.XmlPullParserFactory
import java.io.IOException
import java.io.InputStream
import java.util.ArrayList

class QuestionsXMLparser(xmlStream: InputStream) {
    private val xmlStream: InputStream
    private var qqm: QuizQuestionsModel? = null

    fun getQqm(): QuizQuestionsModel? {
        return this.qqm
    }

    @Throws(XmlPullParserException::class, IOException::class)
    fun parseXML() {
        try {
            Log.v("QuestionsXMLparser", "Entered -> parseXML")
            val pf = XmlPullParserFactory.newInstance()
            val xpp = pf.newPullParser()
            xpp.setFeature(XmlPullParser.FEATURE_PROCESS_NAMESPACES, false)
            xpp.setInput(xmlStream, null)
            xpp.nextTag()
            Log.v("QuestionsXMLparser", "Before processXML method.")
            processXML(xpp)
        } finally {
            Log.v("QuestionsXMLparser", "Closing InputStream.")
            xmlStream.close()
        }
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun processXML(parser: XmlPullParser) {
        Log.v("QuestionsXMLparser", "Entered -> processXML")
        qqm = QuizQuestionsModel()
        var question: QuizQuestion? = null
        val questions: MutableList<QuizQuestion?> = ArrayList()
        parser.require(XmlPullParser.START_TAG, ns, "quiz")
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            val name = parser.name
            if (name == "question") {
                question = parseQuestion(parser)
                Log.v("QuestionsXMLparser", "Question text retrieved: " + question.questionText)
            } else {
                skip(parser)
            }
            questions.add(question)
        }
        qqm!!.questions = questions
        Log.v("QuestionsXMLparser", "Questions parsed: " + qqm!!.questions.size)
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun parseQuestion(parser: XmlPullParser): QuizQuestion {
        var questionId = 0
        var questionText = ""
        var answers: MutableList<QuizAnswer?> = ArrayList()
        parser.require(XmlPullParser.START_TAG, ns, "question")
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                "question_id" -> {
                    questionId = readTagContents(parser, "question_id").toInt()
                }
                "question_text" -> {
                    questionText = readTagContents(parser, "question_text")
                }
                "answers" -> {
                    answers = parseAnswers(parser)
                }
                else -> {
                    skip(parser)
                }
            }
        }
        return QuizQuestion(questionId, questionText, answers)
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun parseAnswers(parser: XmlPullParser): MutableList<QuizAnswer?> {
        val answers: MutableList<QuizAnswer?> = ArrayList()
        parser.require(XmlPullParser.START_TAG, ns, "answers")
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            val name = parser.name
            if (name == "answer") {
                answers.add(parseAnswer(parser))
            } else {
                skip(parser)
            }
        }
        return answers
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun parseAnswer(parser: XmlPullParser): QuizAnswer {
        var answerId = 0
        var answerText = ""
        var answerValid = false
        parser.require(XmlPullParser.START_TAG, ns, "answer")
        while (parser.next() != XmlPullParser.END_TAG) {
            if (parser.eventType != XmlPullParser.START_TAG) {
                continue
            }
            when (parser.name) {
                "answer_id" -> {
                    answerId = readTagContents(parser, "answer_id").toInt()
                }
                "answer_text" -> {
                    answerText = readTagContents(parser, "answer_text")
                }
                "answer_valid" -> {
                    answerValid = readTagContents(parser, "answer_valid").toBoolean()
                }
                else -> {
                    skip(parser)
                }
            }
        }
        return QuizAnswer(answerId, answerText, answerValid)
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun readTagContents(parser: XmlPullParser, tagName: String): String {
        var result = ""
        parser.require(XmlPullParser.START_TAG, ns, tagName)
        if (parser.next() == XmlPullParser.TEXT) {
            result = parser.text
            parser.nextTag()
        }
        return result
    }

    @Throws(XmlPullParserException::class, IOException::class)
    private fun skip(parser: XmlPullParser) {
        check(parser.eventType == XmlPullParser.START_TAG)
        var depth = 1
        while (depth != 0) {
            when (parser.next()) {
                XmlPullParser.END_TAG -> depth--
                XmlPullParser.START_TAG -> depth++
            }
        }
    }

    companion object {
        private val ns: String? = null
    }

    init {
        Log.v("QuestionsXMLparser", "Entered -> QuestionsXMLparser constructor.")
        this.xmlStream = xmlStream
    }
}